fergus_emi_ea-2021-04-ndvi-automation.py

```diff
@@ -0,0 +1,758 @@
+    calculate NDVI and mask using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_paths : string
+        A path to the landsat scene of interest
+    all_bands : list
+        A list containing two xarray objects for landsat bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to crop the raster data using rasterio mask().
+    pixel_qa_path: string
+        A path to a pixel qa tif file.
+    vals: list
+        A list of values needed to create the cloud mask
+
+
+    Returns
+    -----------
+    ndvi_crop : Xarray Dataset
+        a cropped and masked xarray object containing NDVI values
+    """
+    # For loop using open_clean_bands function
+    all_bands = []
+    for aband in band_paths:
+        cleaned_band = open_clean_bands(band_path=aband,
+                                        crop_extent=crop_bound,
+                                        valid_range=(0, 10000))
+        all_bands.append(cleaned_band)
+
+    crop_json = crop_bound.geometry
+
+    # Open and clip qa layer
+    pixel_qa = rxr.open_rasterio(pixel_qa_path[0], masked=True).rio.clip(crop_json,
+                                                                         from_disk=True).squeeze()
+
+    # Calculate NDVI
+    ndvi_xr = (all_bands[1]-all_bands[0]) / (all_bands[1]+all_bands[0])
+
+    # Apply cloud mask to NDVI
+    ndvi_mask = ndvi_xr.where(~pixel_qa.isin(vals))
+
+    return ndvi_mask
+
+
+###############
+#  FUNCTION to create a pandas dataframe from NDVI and grab site and date from file path
+
+
+def mean_ndvi_df(ndvi_mask,
+                 dir_path,
+                 file_path):
+    """Calculate the mean NDVI for a single site and scene with 
+    NDVI calculated and clouds masked out
+
+    Parameters
+    -----------
+    ndvi_mask : Xarray Dataset
+        A cropped and masked xarray object containing NDVI values.
+    dir_path: string 
+        Path to the landsat site folders to grab the site name.
+    file_path: string
+        The name of a individual landsat scene file to grab the date.
+
+
+    Returns
+    -----------
+    mean_ndvi_df : Pandas dataframe
+        a dataframe of mean NDVI by site name and indexed on the date
+    """
+
+    # Calculate mean NDVI for the scene
+    mean_ndvi = np.nanmean(ndvi_mask)
+
+    # Grab site name from file path
+    site = os.path.basename(os.path.normpath(dir_path))
+
+    # Grab date from file name
+    file_name = os.path.basename(file_path)
+    date = file_name[10:18]
+
+    # Create pandas dataframe
+    data = {'date': [date],
+            'site': [site],
+            'mean_ndvi': [mean_ndvi]}
+    mean_ndvi_df = pd.DataFrame(data, columns=['date', 'site', 'mean_ndvi'])
+    mean_ndvi_df['date'] = pd.to_datetime(mean_ndvi_df['date'])  # date format
+    mean_ndvi_df = mean_ndvi_df.set_index('date')  # index dataframe on date
+
+    return mean_ndvi_df
+
+
+# Note: "from_disk" needs updated rioxarray package
+
+
+# In[4]:
+
+
```

Comment:
These comments and hints can be cleaned up once you complete the task

Reply:
Thanks, Korrina for the helpful feedback! I agree that resampling by month was not necessary and may explain the weird x-axis in the plot. It's something I should clean up down the road. And I appreciate you catching the extra libraries that were not needed to run the notebook. Thanks for reviewing my code!